docs/news/posts/2024-11-15-aiidalab-mvc.md

```diff
@@ -0,0 +1,322 @@
+
+```python
+
+import traitlets as tl
+
+class Controller:
+    def __init__(self, model, view):
+        self._model = model
+        self._view = view
+
+        self._set_up_observations()
+
+    def _set_up_observations(self):
+        self._model.observe(
+            self._on_some_change,
+            "something",
+        )
+        tl.dlink(
+            (self._model, "options"),
+            (self._view.dropdown, "options")
+        )
+        tl.link(
+            (self._model, "selected"),
+            (self._view.dropdown, "value")
+        )
+        tl.link(
+            (self._model, "label"),
+            (self._view.textbox, "value")
+        )
+        self._view.button.on_click(self._on_button_click)
+
+    def _on_some_change(self, change):
+        self._do_something(change["new"])
+
+    def _on_button_click(self, _): ...
+
+    def _do_something(self, parameter): ...
+```
+
+The above `Controller` class handles observations in three ways:
+
+1. Observing `Model` changes through the model's `observe` method provided by the `HasTraits` base class
+2. Linking the UI components of the `View` with the data of the `Model` using `traitlets`:
+   1. `dlink` for one-way directional links to receive data from the `Model`, and
+   2. `link` for two-way linking, synchronizing the `View` and `Model`
+3. Observing button clicks using `ipywidgets` `on_click` API for button widgets
+
+For 1 and 3, dedicated `_on_event` methods are a good practice for readability, each encapsulating one or more operations (methods) to perform on either the `View` or the `Model`.
+
+## The model network
+
+The models of the various MVC components form the backbone of the app. To connect the models, once can choose one or both of the following patterns.
+
+### The Observer pattern
+
+In the observer pattern, traits of one MVC `Model` are linked to a copy in another. When the source trait changes, the other is synchronized. The `Controller` of the latter `Model` observes the change and update its local MVC component, thus facilitating the flow of data in the app.
+
+### The Mediator pattern
+
+On can also choose to opt for more granular control. In the Mediator pattern, traits of one `MVC` pattern are observed locally by its `Controller`, with the `Controller` subsequently taking action on another `Model`. In such cases, the latter `Model` is often of a dependent MVC component, of which the `Controller` is aware.
+
+---
+
+**Tip 💡** typically, it is not a matter of choosing one or the other, but instead balancing the two patterns to best fit the logic of the app.
+
+## Services and backends
+
+As apps grow, providing additional features may come with changes in core mechanics. In anticipation of such changes, it is often best to isolate the core mechanics of an app in one or more dedicated services, to be injected into models on demend.
```

Comment:
```suggestion
As apps grow, providing additional features may come with changes in core mechanics. In anticipation of such changes, it is often best to isolate the core mechanics of an app in one or more dedicated services, to be injected into models on demand:
```

Reply:
Hmm, I wouldn't put a colon at the end there. It's not directly referencing the following snippet.